homeassistant/core.py

```diff
@@ -1218,12 +1218,16 @@ def call(
             self._hass.async_create_task(self._safe_execute(handler, service_call))
             return None
 
+        task = self._hass.loop.create_task(self._execute_service(handler, service_call))
         try:
             async with timeout(limit):
-                await asyncio.shield(self._execute_service(handler, service_call))
+                await asyncio.shield(task)
             return True
         except asyncio.TimeoutError:
             return False
+        except asyncio.CancelledError:
+            task.cancel()
```

Comment:
We can't do this. This is literally the only thing that shield does. This means you're removing shield here.

Reply:
FYI https://github.com/python/cpython/blob/89aa7f0ede1a11c020e83f24394593c577a61509/Lib/asyncio/tasks.py#L839

Reply:
No, I don't think so.

If the `timeout` context manager does not expire before the task finishes the `async_call` returns `True`.

If, however, the `timeout` context manager does expire it does two things. First it tries to cancel the `await` _inside_ the `async with`. It does cancel the `await`, but the `shield` prevents it from canceling the task. It then raises a `TimeoutError` _outside_ the `async with` (not a `CancelledError`) which causes the `async_call` to return `False`. The task, however, continues to run, just like before. Nothing different yet.

Now, in the scenario where the task awaiting `async_call` is cancelled, that causes a `CancelledError` to be raised in `await asyncio.shield(task)`, and yes, that will not cancel the task because it is shielded, but then in the `except asyncio.CancelledError` clause the task is _directly_ cancelled (which we can do now because now we have a handle to it) before re-raising the `CancelledError`.

I've tested all four scenarios:

1. Script completes normally before the timeout.
1. Script completes normally but after the timeout.
1. Script is cancelled before the timeout.
1. Script is cancelled after the timeout.

All work as expected.

Reply:
I guess in summary, yes, I'm effectively removing the shield, but only for when async_call is being cancelled, but I'm not for a timeout -- that stays the same as before.

Reply:
We shouldn't use a shield and then work around what does. It's confusing code and we can't have that in the core. 

Instead, can't we use a [`wait`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) with a timer and a task and pass `return_when=FIRST_COMPLETED`?

This is also a big change. We need to make sure that `websocket_api` and `api` integrations both call their services with a shield to avoid cancellation if the connection is dropped. We also probably need to revisit all places that process webhooks and call scripts to make sure it's shielded. Like, a lot more code needs to be cancellation aware that didn't had to be that before. I suggest we drop this change from this PR. It is going to balloon.

Reply:
FWIW, it's not "working around" the shield. The shield is there for one reason, to shield the task from a cancel caused by the timeout. That is still needed. Rather it's adding handling for a cancellation that comes from outside the block of code rather than from inside it. Also I did at first implement it like you said but it created a much larger, and much less elegant, block of code, because wait is such a primitive there is a lot that needed to be duplicated that `timeout` and `shield` already did. In the end, IMHO, it was much more confusing. So I still disagree, but I guess I don't really care one way or the other, as long as it works.

But, more importantly, the more I work on this the more I'm discovering that this basic approach (to enhancing the script helper and its users) is running into, and breaking (or, at least, tripping over), a _lot_ of assumptions. Everywhere I turn I find something else that needs to be addressed, and unfortunately I'm still very new to this code base. And discussing all these issues is not really practical in this PR forum. I'm more than happy to continue working on this -- I'd really like to see it through -- but it's going to take _way_ more time than I had originally thought (even after assuming it wouldn't be easy.)

From here this can go three ways: we put this PR on hold and I take a side step to work on the basic changes to ServiceRegistry.async_call and all its ramifications (and I've already thought of more than we've discusses so far), ultimately in a separate PR, and then pick this one back up later, working both PRs much more slowly and methodically; I throw in the towel and you or someone else takes up the charge, using whatever you like from what I started; or we completely scrap this whole approach (reverting the previous PR) until you and/or someone else can figure out how to start it again a different way. Your choice.

Reply:
I think that you're on the right track here with scripts etc. I think that cancellation of a subscript is not a show-stopper issue. It's annoying that it will only cancel after the current script is done, but that's still an improvement over today.

The shield is not just there to cancel out the timeout. It's to keep it running, no matter if the `async_call` is canceled, either through a timeout or because the caller canceled. Because this code has been there since day 1, none of the service handling code deals with cancellations. 

I think that we should be able to merge this PR without removing the shield. The cancellation will just make the current script finish. We can improve on that later. I'm happy to help and think along etc. I think that it's a great improvement. I don't think that it's necessary to work in parallel or throw in the towel.

Reply:
Ok, thanks, that's encouraging.

Yes, I see what you're saying about service handlers not handling cancellation. (I had to add that to the `script.turn_on` service handler.) I was only thinking from the other direction -- callers of `async_call`. Good point.

I'll continue slugging away. I'll probably go silent for a while as I try to work through the issues and come up with reasonable approaches to solving them. Besides the changes to `async_call`, there are still a few cleanup things to accomplish on this PR assuming we'll merge it without those changes. If I need help should I keep the discussion going here, or would you recommend another channel (discord, etc.)?

Reply:
This channel works, that way we can discuss it around parts of the code.

Reply:
btw I believe that small PRs keep things manageable. Even if we don't expose new functionality yet, we can already merge refactor PRs like this one, that shouldn't impact normal operational behavior.

Reply:
Okey dokey. About to push two commits. The first backs out the last changes to homeassistant/core.py, and the second includes the minor fixes I mentioned. It also fairly overhauled tests/helpers/test_script.py, using Parametrize, using/adding common test functions, and I found a way to mock timeout so the tests don't have to wait for real time to elapse.